fix: #23915 podman build is not parsing sbom command line arguments

aguidirh opened the pull request:
# Issue Description
SBOM flags are not respected while podman build command.

At the same time buildah build command works as expected.

Fixes #23915 

## Steps to reproduce the issue
With the following Containerfile:
```
FROM ubuntu:22.04
WORKDIR /app
```

Run the following podman build:

```
podman build -t sbom-img --sbom=trivy-spdx \
        --sbom-image-output=/app/sbom-spdx.json \
        --sbom-output=sbom-spdx.json \
        --sbom-scanner-image=ghcr.io/aquasecurity/trivy \
        --sbom-scanner-command="trivy filesystem -q {ROOTFS} --format spdx-json --output {OUTPUT}" \
        --sbom-scanner-command="trivy filesystem -q {CONTEXT} --format spdx-json --output {OUTPUT}" \
        --sbom-merge-strategy=merge-spdx-by-package-name-and-versioninfo \
        -f Containerfile
```

Create a container with the image built in the previous step and check if the file sbom-spdx.json is inside of the container as requested:

```
podman run -it --rm sbom-img ls -al
```

Expected result:
```
drwxr-xr-x. 1 root root     28 Mar 21 13:06 .
dr-xr-xr-x. 1 root root     12 Mar 21 13:17 ..
-rw-r--r--. 1 root root 147729 Mar 21 13:06 sbom-spdx.json
```

Actual results:
```
drwxr-xr-x. 1 root root     28 Mar 21 13:06 .
dr-xr-xr-x. 1 root root     12 Mar 21 13:17 ..
```

Running the steps above with the code from this PR shows the expected result (the same as when using buildah), while running with the code from the main branch shows the actual result (with the bug).


#### Does this PR introduce a user-facing change?

```release-note
None
```

baude commented:
please sign your commits https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs

baude commented:
@nalind ptal

baude commented:
and you need to add a release note ... this is cli and is an outward change.

aguidirh commented:
Hi @baude,

> please sign your commits https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs

I believe the [commit](https://github.com/containers/podman/pull/25647/commits) is signed.

> and you need to add a release note

I did not add a release note because in the template it was saying to add it only in case of user-facing change, there is no user-facing change, only a bug fix for the flags that were already there.

nalind commented:
This looks exactly what `github.com/containers/buildah/pkg/cli.GenBuildOptions()` does, so I expect it's going to handle the options correctly, so LGTM.
But if podman isn't calling that function, I have to wonder why we went to the trouble of exporting it there.

aguidirh commented:
/retest

aguidirh commented:
Hi @nalind,

I added the test requested by @baude. 

They are passing locally, I am not sure why some tests are failing in the PR.

Could you please /retest them?

nalind commented:
/ok-to-test

nalind commented:
/retest

baude commented:
the test `[FAIL] Podman build [It] podman build with sbom flags` is failing here.

baude commented on test/e2e/build_test.go:
```diff
@@ -972,4 +972,19 @@ RUN ls /dev/test1`, CITEST_IMAGE)
+		session.WaitWithDefaultTimeout()
+		Expect(session).Should(ExitCleanly())
+
+		session = podmanTest.Podman([]string{"run", "--rm", "sbom-img", "ls", "/tmp"})
```
same as above PodmanExitCleanly

aguidirh replied:
Great suggestion, thanks. 

Done.

baude commented:
also note the failures are all remote tests

aguidirh commented on test/e2e/build_test.go:
```diff
@@ -972,4 +972,19 @@ RUN ls /dev/test1`, CITEST_IMAGE)
 		Expect(session).Should(ExitWithError(1, `building at STEP "RUN --mount=type=cache,target=/test,z cat /test/world": while running runtime: exit status 1`))
 	})
+	It("podman build with sbom flags", func() {
+		podmanTest.AddImageToRWStore(ALPINE)
```
Great suggestion, thanks. 

Done.

aguidirh commented:
> also note the failures are all remote tests

I will have a look why they are failing for the remote tests and fix it.

l0rd commented:
Great job @aguidirh , tests are passing. If you don't mind, it would be helpful to add some remote tests (see [pkg/machine/e2e](https://github.com/containers/podman/blob/main/pkg/machine/e2e)). Also, please squash your commits as CI tests were not passing for the first one.

aguidirh commented:
> Great job @aguidirh , tests are passing. If you don't mind, it would be helpful to add some remote tests (see [pkg/machine/e2e](https://github.com/containers/podman/blob/main/pkg/machine/e2e)). Also, please squash your commits as CI tests were not passing for the first one.

Thanks @l0rd. 

In test/e2e/build_test.go [L977-997](https://github.com/containers/podman/pull/25647/files#diff-8b4d34c763f953caff8c6747e966aa1a7e869411747cd000370918887ba5b124R977-R997) I added `podman build with sbom flags` which I believe runs as `make localintegration` and also as remote test `make remoteintegration`

Is this kind of remote test are you asking for? Or is it other type of remote test? 

I squashed the commits, but I believe we have some flaky tests.

aguidirh commented:
/hold 

TODO: discover why build with sbom flags are not working for podman machine `localmachine` on `basic_test.go` L[326-327](https://github.com/containers/podman/pull/25647/files#diff-0703eb2944434bdd3c13ee3d30076c101479b02fa257b6935e36914b06d8817eR326-R327). 

The same test works fine for `localintegration` and  `remoteintegration` on `test/e2e/build_test.go` L[995-996](https://github.com/containers/podman/pull/25647/files#diff-8b4d34c763f953caff8c6747e966aa1a7e869411747cd000370918887ba5b124R995-R996).

aguidirh commented:
@mheon since the machine tests will need to wait for the new vm to be built based on the server code I'm sending in this PR, is that ok if I comment the entire machine test ? 

In this way we can check if the problem is that my tests bumped over the limit. 

What do you think?

l0rd commented on pkg/machine/e2e/basic_test.go:
```diff
@@ -287,6 +287,46 @@ var _ = Describe("run basic podman commands", func() {
 		Expect(build.outputToString()).To(ContainSubstring(name))
 	})
 
+	It("podman build with sbom flags", func() {
+
+		name := randomString()
+		i := new(initMachine)
+		session, err := mb.setName(name).setCmd(i.withImage(mb.imagePath).withNow()).run()
+		Expect(err).ToNot(HaveOccurred())
+		Expect(session).To(Exit(0))
```
I am sorry, I am answering late but for some reason I didn't get any notification. I think I asked that to cover both daemonless and daemonful scenarios.

Luap99 replied:
Ack, the test in test/e2e and test/system already get executed locally and via remote client. It is just that client and server are on the same linux system so it is not suited to catch macos or windows client problems but for things like this the e2e case is good enough.

We need to be careful to not add to many machine tests due its slowness/flakiness so I prefer if this case is dropped.

l0rd replied:
@Luap99 Makes sense. @aguidirh Apologies for asking to add an unnecessary test 🤷‍♂️

aguidirh replied:
Don't worry @l0rd, it was good for me to learn more about how the machine tests work.

I removed them as suggested by @Luap99.

aguidirh commented:
@Luap99 all tests are passing now, could you please take a look to see if there is anything missing before merging this PR?
